File: educational-core/src/com/jetbrains/edu/learning/StudyTaskManager.java
package com.jetbrains.edu.learning;

import com.intellij.ide.lightEdit.LightEdit;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.components.*;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Computable;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.util.messages.Topic;
import com.intellij.util.xmlb.annotations.Transient;
import com.jetbrains.edu.learning.authorContentsStorage.AuthorContentsStorageUtilsKt;
import com.jetbrains.edu.learning.authorContentsStorage.zip.UpdatableZipAuthorContentsStorage;
import com.jetbrains.edu.learning.courseFormat.Course;
import com.jetbrains.edu.learning.courseFormat.authorContentsStorage.AuthorContentsStorage;
import com.jetbrains.edu.learning.courseFormat.ext.CourseExt;
import com.jetbrains.edu.learning.yaml.YamlDeepLoader;
import com.jetbrains.edu.learning.yaml.YamlFormatSettings;
import com.jetbrains.edu.learning.yaml.YamlFormatSynchronizer;
import kotlin.Unit;
import org.jdom.Element;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import static com.jetbrains.edu.learning.OpenApiExtKt.getCourseDir;

/**
 * Implementation of class which contains all the information
 * about study in context of current project
 */
@State(name = "StudySettings", storages = @Storage(value = "study_project.xml", roamingType = RoamingType.DISABLED))
public class StudyTaskManager implements PersistentStateComponent<Element>, DumbAware, Disposable {
  public static final Topic<CourseSetListener> COURSE_SET = Topic.create("Edu.courseSet", CourseSetListener.class);
  private volatile boolean courseLoadedWithError = false;

  @Transient
  private Course myCourse;

  @Transient @Nullable private final Project myProject;

  /**
   * This is the author contents storage used to store all the edu files contents for the current course.
   * It is updatable, that is, if the course is updated by the student, this storage also updates.
   *
   * This storage is needed only in the student mode.
   * It should not be used in the course creation mode.
   */
  @Transient @Nullable
  private final UpdatableZipAuthorContentsStorage myAuthorContentsStorage;

  public StudyTaskManager(@Nullable Project project) {
    myProject = project;

    if (myProject != null) {
      VirtualFile courseDir = getCourseDir(project);
      myAuthorContentsStorage = new UpdatableZipAuthorContentsStorage(courseDir);
    } else {
      myAuthorContentsStorage = null;
    }
  }

  public StudyTaskManager() {
    this(null);
  }

  @Transient
  public void setCourse(Course course) {
    myCourse = course;
    if (myProject != null) {
      myProject.getMessageBus().syncPublisher(COURSE_SET).courseSet(course);
    }
  }

  @Nullable
  @Transient
  public Course getCourse() {
    return myCourse;
  }

  @Nullable
  @Override
  public Element getState() {
    return null;
  }

  @Nullable
  @Transient
  public AuthorContentsStorage getAuthorContentsStorage() {
    return myAuthorContentsStorage;
  }

  /**
   * This method saves all current edu file contents to the author contents storage,
   * and reassigns edu file contents to point to this storage.
   * We may not reassign edu file contents to point on the storage, but we do it to free up some resources,
   * for example, if some file contents are pointing into memory.
   */
  @Transient
  public void updateAuthorContentsStorageAndTaskFileContents() {
    if (myAuthorContentsStorage != null)
      myAuthorContentsStorage.update(myCourse);

    CourseExt.visitEduFiles(myCourse, eduFile -> {
      eduFile.setContents(AuthorContentsStorageUtilsKt.fileContentsFromProjectAuthorContentsStorage(eduFile));
      return Unit.INSTANCE;
    });
  }

  @Override
  public void loadState(@NotNull Element state) {
  }

  @Override
  public void dispose() { }

  public static StudyTaskManager getInstance(@NotNull final Project project) {
    StudyTaskManager manager = ServiceManager.getService(project, StudyTaskManager.class);
    if (!project.isDefault() &&
        !LightEdit.owns(project) &&
        manager != null &&
        manager.getCourse() == null &&
        YamlFormatSettings.isEduYamlProject(project) &&
        !manager.courseLoadedWithError) {
      Course course = ApplicationManager.getApplication().runReadAction((Computable<Course>)() -> YamlDeepLoader.loadCourse(project));
      manager.courseLoadedWithError = course == null;
      if (course != null) {
        manager.setCourse(course);
      }
      YamlFormatSynchronizer.startSynchronization(project);
    }
    return manager;
  }
}

File: educational-core/src/com/jetbrains/edu/learning/StudyTaskManager.kt
package com.jetbrains.edu.learning

import com.intellij.ide.lightEdit.LightEdit
import com.intellij.openapi.Disposable
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.components.service
import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Computable
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.util.messages.Topic
import com.intellij.util.xmlb.annotations.Transient
import com.jetbrains.edu.learning.authorContentsStorage.*
import com.jetbrains.edu.learning.courseFormat.*
import com.jetbrains.edu.learning.courseFormat.ext.getPathInCourse
import com.jetbrains.edu.learning.courseFormat.ext.visitEduFiles
import com.jetbrains.edu.learning.yaml.YamlDeepLoader.loadCourse
import com.jetbrains.edu.learning.yaml.YamlFormatSettings.isEduYamlProject
import com.jetbrains.edu.learning.yaml.YamlFormatSynchronizer.startSynchronization
import java.nio.file.Files
import java.nio.file.Path

/**
 * Implementation of class which contains all the information about study in context of current project
 */
@Service(Service.Level.PROJECT)
class StudyTaskManager(private val project: Project) : DumbAware, Disposable {
  @Volatile
  private var courseLoadedWithError = false

  private var _course: Course? = null

  /**
   * This is the author contents storage used to store all the edu files contents for the current course.
   * It is updatable, that is, if the course is updated by the student, this storage also updates.
   * This storage is needed only in the student mode.
   * It should not be used in the course creation mode.
   */
  @Transient
  lateinit var authorContentsStorage: SQLiteContentsStorage
    private set

  var course: Course?
    get() = _course
    set(course) {
      _course = course

      if (_course?.isStudy == true) {
        if (!this::authorContentsStorage.isInitialized) {
          authorContentsStorage = SQLiteContentsStorage.openOrCreateDB(getZipPathForProjectDirectory(project))
          Disposer.register(this, authorContentsStorage)
        }
        updateCourseFileContentsHolders()
      }

      course?.apply {
        project.messageBus.syncPublisher(COURSE_SET).courseSet(this)
      }
    }

  override fun dispose() {}

  /**
   * Should be called each time the course is updated
   */
  private fun updateCourseFileContentsHolders() {
    _course?.visitEduFiles { eduFile ->
      val contents = eduFile.contents
      if (contents is SQLiteContents && contents.storage == authorContentsStorage) {
        return@visitEduFiles
      }

      val path = if (eduFile is TaskFile) {
        eduFile.task.getPathInCourse() + "/"
      }
      else {
        ""
      } + eduFile.name

      eduFile.contents = when (contents) {
        is BinaryContents -> {
          authorContentsStorage.put(path, contents.bytes)
          SQLiteBinaryContents(authorContentsStorage, path)
        }
        is TextualContents -> {
          authorContentsStorage.put(path, contents.text)
          SQLiteTextualContents(authorContentsStorage, path)
        }
        is UndeterminedContents -> {
          authorContentsStorage.put(path, contents.textualRepresentation)
          SQLiteUndeterminedContents(authorContentsStorage, path)
        }
      }
    }
  }

  companion object {
    val COURSE_SET = Topic.create("Edu.courseSet", CourseSetListener::class.java)

    fun getInstance(project: Project): StudyTaskManager {
      val manager = project.service<StudyTaskManager>()
      if (!project.isDefault && !LightEdit.owns(project) && manager.course == null
          && project.isEduYamlProject() && !manager.courseLoadedWithError) {
        val course = ApplicationManager.getApplication().runReadAction(Computable { loadCourse(project) })
        manager.courseLoadedWithError = course == null
        if (course != null) {
          manager.course = course
        }
        startSynchronization(project)
      }
      return manager
    }

    const val COURSE_AUTHOR_CONTENTS_FILE = ".author_contents_storage_db"
    private val courseDir2zipDir = mutableMapOf<String, Path>()

    private fun getZipPathForProjectDirectory(project: Project): Path {
      val courseDir = project.courseDir
      val courseDirPath = LocalFileSystem.getInstance().getNioPath(courseDir) ?: // null, if the test mode
        courseDir2zipDir.computeIfAbsent(courseDir.path) {
          Files.createTempDirectory("boundAuthorContentStorage")
        }

      return courseDirPath.resolve(COURSE_AUTHOR_CONTENTS_FILE)
    }
  }
}
